### PR TITLE
Respect uncontrolled inner blocks on Navigation block in editor and front of site

### PR DESCRIPTION
**Pull request by getdave**
## What?

**This change means that Themes **should not be including inner blocks in their Navigation blocks** in their templates/parts. The only exception to this would be if they wanted to opt out of some of the current / future menu picking optimisations and automations provided by the block (e.g. carrying across menus on Theme switch).**

----
This PR updates the Navigation block to prefer and _preserve_ uncontrolled inner blocks over all other types of Navigation "content" that a Navigation block can display.

This means that if there are any uncontrolled inner blocks present the block will:

- **_not_** attempt to find automatically assign the most appropriate Navigation Menu (`wp_navigation`).
- **_not_** replace the inner blocks with a Navigation Menu, even if a `ref` is present. 

**Important:** The major change with this approach is that if [any themes use the Page List block as a pseudo "placeholder" in their Navigation block](https://github.com/WordPress/twentytwentytwo/blob/27054dbcb95f5d72cd943a0f3318c3ad97599f85/parts/header.html#L9-L11) (as follows):

```
<!-- wp:navigation --><!-- wp:page-list /--><!-- /wp:navigation -->
```

...then that Theme is effectively **opting out of any current/future automatic picking of menus** (e.g. on Theme switch).

For example, currently if there is a single `wp_navigation` post present on the site then the Nav block will automatically pick and use that menu when the Nav block is inserted. This is useful for scenarios such as Theme switching as the menu will effectively be automatically "carried over" to the new Theme without requiring user intervention. In the future we anticipate this functionality being extended and improved.

## Why?

In https://github.com/WordPress/gutenberg/issues/38291#issuecomment-1171582948 we identified that the block is currently exhibiting different behaviour between the Editor and the front of the site.

This PR normalizes the behaviour so that the block will always respect any uncontrolled inner blocks if they are present.

This allows use cases such as creating Navigation patterns in the editor.

## How?

In the effect which attempts to _automatically_ assign a Navigation Post to the block, we now test for uncontrolled inner blocks. If these are present then we abort the automated behaviour thereby allowing the uncontrolled blocks to be rendered.

Note that if the user makes any changes to the uncontrolled blocks to make them "dirty" then they will still be automatically converted into a `wp_navigation` post (existing behaviour).

### Basic Testing
1. Create a single Navigation (`wp_navigation`) post. You can do this by creating an empty Navigation block and adding some items and saving in the multi-entity saving flow.
2. Validate you have only a single `wp_navigation` post (tip: goto `wp-admin/edit.php?post_type=wp_navigation`).
3. New Post.
4. Toggle code editor mode in the Editor.
5. Paste `<!-- wp:navigation --><!-- wp:page-list /--><!-- /wp:navigation -->`.
6. Exit code editor mode.
7. See the Page List block preserved within the Navigation block.
8. You can check by using List View and/or toggling Code Editor mode again and making sure there is no `ref` attribute and that you can still see in the block markup for the inner blocks.
9. Publish the post and validate the exact same content is output on the front of the site (i.e. a Navigation with a page list).

### Theme Switch testing

- again check you only have a single Navigation Post (see above).
- Install Archeo (note that currently [Archeo uses a Page List as inner blocks](https://github.com/Automattic/themes/blob/8ac496ee802c4cb6973b68772b96a274d56bcf56/archeo/parts/header.html#L5-L7)).
- switch Themes to Archeo
- go to Site Editor
- See that Archeo is showing the Page List block and has not carried across your Navigation items.
- Modify this file in Archeo and remove the inner `<!-- wp:page-list /-->`.
- Reload the Site Editor (simulates a Theme switch).
- See that now the items from your Navigation are correctly carried across and the page list is ignored.
- DO NOT SAVE
- Switch to front of site.
- See that it's rendering the exact same thing.

## Screenshots or screencast 

https://user-images.githubusercontent.com/444434/177537722-d63cb2cd-298d-4c97-9fc9-4a235cf3998d.mov

### CODE REVIEW

**Review by draganescu — APPROVED** [review]
I think this is a good change that makes a lot more sense than the current behavior and helps the development of the block evolve into a more seamless result.

**Comment by getdave** [comment]
@scruffian @MaggieCabrera Do we think we can ship this now or should we wait until we get the "Page List block as a fallback" working in the _editor_?

**Comment by getdave** [comment]
@WordPress/block-themers do you have an opinion on this PR? It effects how themes that use a page list block in their Nav block within templates will behave. Description has full context.

**Comment by mikachan** [comment]
> Do we think we can ship this now or should we wait until we get the "Page List block as a fallback" working in the editor?

My opinion would be to ship this now as I think it's a good improvement to the block. We should highlight the implementation change in release notes, but I don't think it should wait until we have the same behaviour in the editor.

**Comment by getdave** [comment]
TL;DR

- don't merge this PR until https://github.com/WordPress/gutenberg/issues/42563 is done.
- Page List fallback should be throw away in the editor and not-persisted
- block should be considered "empty" until the user actively selects something on interaction with the block

------

Upon reflection I think we should wait on https://github.com/WordPress/gutenberg/issues/42563 before merging this. The reason is that until we have Pages as a part of the Nav block's fallback states then Themes will want to use Page List in their templates in order to match what already happens on the front of the site.

Once https://github.com/WordPress/gutenberg/issues/42563 is done then the block will auto-handle displaying a list of Pages in the Editor as the Nav block's fallback state thereby removing the need to manually utilise a Page List as an inner block.

@draganescu The issue we will have in using a Page List block as a fallback is that it will be an uncontrolled inner block. At which point thanks to _this_ PR the block will always respect those inner blocks. This means that as a user you might insert a Nav block only to have a Page List foisted upon you for ever more.

What we actually want is for the Page List block to act as a fallback/placeholder only. If the user doesn't manually interact with the Nav block then it [the Page List] should be discarded and the block still considered empty.

**Comment by getdave** [comment]
@draganescu what's the deal with this PR now https://github.com/WordPress/gutenberg/pull/42735 is merged?

**Comment by draganescu** [comment]
I think this one can be closed and you to be a co-author of https://github.com/WordPress/gutenberg/pull/42735 as that was based on this one /o'\

**Comment by getdave** [comment]
Closing in favour of https://github.com/WordPress/gutenberg/pull/42735

**Comment by getdave** [comment]
I've suggested we write a dev note for WP 6.1 based on changes that were outlined in this PR which ultimately ended up in [another PR](https://github.com/WordPress/gutenberg/pull/42735#issuecomment-1254739231).
